# 01 Import Libraries and Data

In [40]:
#Import the pandas, NumPy, and os libraries.
import pandas as pd
import numpy as np
import os

In [41]:
# Turn your project folder path into a string as outlined in the Exercise text.
path = r'/Users/peterreadman/Desktop/Python Projects/CareerFoundry/04-2020 Instacart Basket Analysis/'

In [45]:
# import orders.csv data set
df_ords = pd.read_csv(os.path.join(path, '02 Data', '02a Original Data', 'orders.csv'), index_col = False)

In [43]:
# import products.csv data set
df_prods = pd.read_csv(os.path.join(path, '02 Data', '02a Original Data', 'products.csv'), index_col = False)

In [48]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [49]:
df_prods.shape

(49693, 5)

In [47]:
df_ords.shape

(3421083, 7)

# Exercise 4.4 Data Wrangling Procedures

In [6]:
# remove 'eval_set' column from view (it's retained in dataframe)
df_ords = df_ords.drop(columns = ['eval_set'])

In [7]:
# check for missing values (NaN) in 'days_since_prior_order' column
df_ords['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [8]:
# rename column 'orders_dow' to 'orders_day_of_the_week'
df_ords.rename(columns = {'order_dow' : 'ordersday_of_the_week'}, inplace = True)

In [9]:
df_ords.head()

order_id  user_id  order_number  ordersday_of_the_week  order_hour_of_day  \
0   2539329        1             1                      2                  8   
1   2398795        1             2                      3                  7   
2    473747        1             3                      3                 12   
3   2254736        1             4                      4                  7   
4    431534        1             5                      4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [10]:
# rename column 'ordersday_of_the_week' to 'orders_day_of_the_week'
df_ords.rename(columns = {'ordersday_of_the_week' : 'orders_day_of_the_week'}, inplace = True)

In [11]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0   2539329        1             1                       2                  8   
1   2398795        1             2                       3                  7   
2    473747        1             3                       3                 12   
3   2254736        1             4                       4                  7   
4    431534        1             5                       4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [12]:
# change variable data type - e.g. change id numbers from INT to STR to avoid inclusion in stats
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [13]:
df_ords['order_id'].dtype

dtype('O')

In [14]:
# import departments.csv data set
df_dep = pd.read_csv(os.path.join(path, '02 Data', '02a Original Data', 'departments.csv'), index_col = False)

In [15]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [16]:
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [17]:
df_dep_t = df_dep.T

In [18]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [19]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [20]:
# Take the first row of df_dep_t for the header
new_header = df_dep_t.iloc[0]

In [21]:
new_header

0    department
Name: department_id, dtype: object

In [22]:
df_dep_t_new = df_dep_t[1:]

In [23]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [24]:
df_dep_t_new.columns = new_header #set the header row as the df header

In [25]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

# Data Dictionaires

In [26]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [27]:
data_dict = df_dep_t_new.to_dict('index')

In [28]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [29]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [30]:
print(data_dict.get('19'))

{'department': 'snacks'}


# Subsetting

In [31]:
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [32]:
df_snacks = df_prods[df_prods['department_id']==19]

In [33]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [34]:
df_snacks_2 = df_prods.loc[df_prods['department_id'] == 19]

In [35]:
df_snacks_2.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [36]:
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]

In [37]:
df_snacks_3.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

# TASK 4.4 

# Step 1
data wrangling above

# Step 2
Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.


In [38]:
# change user_id to a string

df_ords['user_id'] = df_ords['user_id'].astype('str')


In [39]:
# check 'user_id' is a string by using dtype()
df_ords['user_id'].dtype

dtype('O')

# Step 3
Look for a variable in your df_ords dataframe with an unintuitive name
and change its name without overwriting the data frame.

In [94]:
# rename column 'days_since_prior_order' to 'days_since_last_order'
df_ords.rename(columns = {'days_since_prior_order' : 'days_since_last_order'}, inplace = True)

In [93]:
df_ords.head()

order_id user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0  2539329       1             1                       2                  8   
1  2398795       1             2                       3                  7   
2   473747       1             3                       3                 12   
3  2254736       1             4                       4                  7   
4   431534       1             5                       4                 15   

   days_since_last_order  
0                    NaN  
1                   15.0  
2                   21.0  
3                   29.0  
4                   28.0

# Step 4
Your client wants to know what the busiest hour is for placing orders.
Find the frequency of the corresponding variable and share your findings.

In [98]:
df_ords['order_hour_of_day'].value_counts()

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

**Answer**<br>The busiest hour for placing orders is 10am (with 288,418 orders placed at 'order_hour_of_day' = 10)

# Step 5
Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.


In [103]:
print(data_dict.get('4'))

{'department': 'produce'}


**Answer**<br>Products with 'department_id' of 4 are in the **produce** department

# Step 6
The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.


In [108]:
# call up the data dictionary to check if there's a 'breakfast' department_id to filter
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [109]:
# there is! department_id = 14 = 'breakfast'
# create subset containing only 'breakfast' items

df_breakfast = df_prods[df_prods['department_id']==14]

In [110]:
# check the subset is created
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

# Step 7
They’d also like to see details about customers who might
departments: alcohol (5), deli (20), beverages(7), and meat/seafood (12). You’ll need to present this subset to your client.

In [111]:
# use the .loc function to look into a list

df_dinner_parties = df_prods.loc[df_prods['department_id'].isin([5,20,7,12])]

In [112]:
df_dinner_parties

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

# Step 8
It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

**Answer**<br>The new 'df_dinner_parties' dataframe has **7650 rows**

# Step 9
Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [113]:
# create a subset for user_id = 1

df_user_id_1 = df_ords.loc[df_ords['user_id'] == 1]

In [114]:
df_user_id_1

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_the_week, order_hour_of_day, days_since_last_order]
Index: []

**Answer**<br>It looks like user_id 1 has zero orders

# Export
**df_ords** dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.<br>**df_dep_t_new** dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format.

In [116]:
df_ords.to_csv(os.path.join(path, '02 Data', '02b Prepared Data', 'orders_wrangled.csv'))